src/main/scala/is/hail/expr/FunctionRegistry.scala

```diff
@@ -2914,4 +2923,9 @@ object FunctionRegistry {
         s[-i:-j] == s[s.length - i, s.length - j]
     """, "i" -> "Starting index of the slice.", "j" -> "End index of the slice (not included in result)."
   )
+
+  registerDependentCode("str", { (tt: TypeTag) => (v: Code[AnyRef]) => CM.invokePrimitive1(tt.xs(0).str)(v) },
```

Comment:
I don't think you need to pass TypeTag for any of the applications we're considering, including this one.  You can write something like:

```
  registerDependentCode("str", { () =>
    val t = TTRealizable.t
    (v: Code[AnyRef]) => CM.invokePrimitive1(t.str)(v)
  }, ...)
```

This seems easier than digging desired fields out of the TypeTag.  We just need the thunk to be run while the matched values are still present in the variables.

Reply:
I do not like referencing the type variable because it trades explicit data flow for implicit data flow.

Digging sounds to me like repeated field accesses, which isn't the case here. TypeTag is an array of types (plus some unification methods). The only more parsimonious implementation that I can imagine (while using explicit data flow) is:

```
case class UnaryDependentCode[T,U](..., code: Type => ...) {
  override def typeapply(tt: TypeTag): Fun =
    new UnaryFunCode[T, U](retType: Type, code(tt.xs(0)))
}
```

Reply:
> I do not like referencing the type variable because it trades explicit data flow for implicit data flow.

I'm not sure what you mean by explicit dataflow.  I would have said exactly the opposite.  In both use cases, the value in the variable IS the value I want.

Reply:
Explicit in the sense that values are explicitly passed into the method, rather than implicitly passed via external state.